### PR TITLE
Hackable distributed filesystem reader and writer

### PR DESCRIPTION
**Pull request by hanzhi713**
I propose some changes so that the `FileSystemReader` and `FileSystemWriter` can be used on other file systems. User only needs to provide `path` as a subclass of `Path` that overrides the necessary interfaces. 

For example, one can utilize `tf.io.gfile` to implement an interface to save to or load from HDFS. The following code snippet shows a working implementation. 

```python
from pathlib import Path
import tensorflow as tf


class GFileWrapper(tf.io.gfile.GFile):
    def __init__(self, path, mode="r") -> None:
        super().__init__(path, mode)
        
    def write(self, data):
        return super().write(bytes(data))
    
    # a not quite efficient readinto, but it works
    def readinto(self, buffer):
        # read up to buffer's length
        data = self.read(len(buffer))
        length = len(data)
        buffer[:length] = data
        return length


class HdfsPath(type(Path())):
    def __new__(cls, *pathsegments):
        return super().__new__(cls, *pathsegments)
    
    @staticmethod
    def _fix_path(path):
        path = str(path)
        if path.startswith("hdfs:/") and not path.startswith("hdfs://"):
          path = path.replace("hdfs:/", "hdfs://")
        return path

    def open(self, mode="r", *args, **kwargs):
        return GFileWrapper(HdfsPath._fix_path(self), mode=mode)
    
    def mkdir(self, **kwargs) -> None:
        return tf.io.gfile.makedirs(HdfsPath._fix_path(self))
    
    def rename(self, target):
        return tf.io.gfile.rename(HdfsPath._fix_path(self), HdfsPath._fix_path(target))
```

```python
writer = FileSystemWriter(HdfsPath("hdfs://..."), sync_files=False)
reader = FileSystemReader(HdfsPath("hdfs://..."))
```

### CODE REVIEW

**Comment by hanzhi713** [comment]
@mrshenli @zhaojuanmao @rohan-varma @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @kiukchung @d4l3k 
Could you please take a look at this PR when you have time? I'm hoping to get this merged before the 2.1 release.

File: torch/distributed/checkpoint/filesystem.py
Comment (fduwjj) [inline torch/distributed/checkpoint/filesystem.py:451]: nit: 
```
maybe self.path = path if isinstance(path, Path) else Path(path)
```

**Review by fduwjj — APPROVED** [review]
LGTM

File: torch/distributed/checkpoint/filesystem.py
Comment (hanzhi713) [reply torch/distributed/checkpoint/filesystem.py:451]: I'm going to keep this unchanged unless pytorch guideline prefers one liner.

**Comment by kit1980** [comment]
@hanzhi713 @fduwjj Should we merge this?

**Comment by hanzhi713** [comment]
> @hanzhi713 @fduwjj Should we merge this?

I'm hoping to. Just need maintainers to approve.

**Review by fduwjj — APPROVED** [review]

**Comment by fduwjj** [comment]
Well, I think so. But how comes after accepting, this PR can still not be merged? @kit1980

**Comment by kit1980** [comment]
> Well, I think so. But how comes after accepting, this PR can still not be merged?

@fduwjj What to you mean? We merge by posting "@pytorchbot merge" comment.

**Comment by hanzhi713** [comment]
@kit1980 What's going on here 😅 ?

**Comment by kit1980** [comment]
@hanzhi713 it was timeout

**Pull request closed without merging**
